### PR TITLE
PACT-808 Create unit test suite for LoginController

### PR DESCRIPTION
**Pull request by userjd**

### CODE REVIEW

File: test/controllers/LoginControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+class LoginControllerSpec extends BaseControllerSpec {
+
+  "LoginController GET" should {
+
+    "render the login page from a new instance of controller" in {
+
+      val response = getLoginController.view().apply(CSRFTokenHelper.addCSRFToken(FakeRequest(GET, "/login")))
+
+      status(response) mustBe OK
+      val document = asDocument(response)
```
Comment (tna-erasmos) [inline test/controllers/LoginControllerSpec.scala:42]: We shouldn't have any assertions on the contents; these belong in the view specs, as well as the integration tests.

File: test/controllers/LoginControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      document must haveLegend("login.heading")
+
+    }
+
+    "render the login page from the application" in {
```
Comment (tna-erasmos) [inline test/controllers/LoginControllerSpec.scala:47]: This is identical to the one above. 

We're no longer using the application, but rather the controller, so the title is wrong, as well. 

I'd recommend deleting this scenario.

File: test/controllers/LoginControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+  "LoginController POST" should {
+    "redirect to result page given valid username and password from a new instance of controller" in {
+
+      val response = getLoginController
+        .submit()
+        .apply(
+          CSRFTokenHelper
+            .addCSRFToken(
+              FakeRequest(POST, "/login").withFormUrlEncodedBody("username" -> "1234", "password" -> "1234")
```
Comment (tna-erasmos) [inline test/controllers/LoginControllerSpec.scala:67]: The actual path is ignored; you might as well use a blank.

File: test/controllers/LoginControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+      val response = getLoginController
+        .submit()
+        .apply(
+          CSRFTokenHelper
+            .addCSRFToken(
+              FakeRequest(POST, "/login").withFormUrlEncodedBody("username" -> "1234", "password" -> "12345")
+            )
+        )
+      status(response) mustBe BAD_REQUEST
```
Comment (tna-erasmos) [inline test/controllers/LoginControllerSpec.scala:84]: Shouldn't we assert the redirection location?

File: test/controllers/LoginControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+          CSRFTokenHelper
+            .addCSRFToken(
+              FakeRequest(POST, "/login").withFormUrlEncodedBody("username" -> "1234", "password" -> "1234")
+            )
+        )
+      status(response) mustBe SEE_OTHER
+      session(response).get(SessionKeys.token) must not be empty
+    }
+
+    "redirect to login page given valid username and invalid password from a new instance of controller" in {
```
Comment (tna-erasmos) [inline test/controllers/LoginControllerSpec.scala:74]: It's not a redirect, but rather a re-presentation of the same page.

File: test/controllers/LoginControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+          CSRFTokenHelper
+            .addCSRFToken(
+              FakeRequest(POST, "/login").withFormUrlEncodedBody("username" -> "1234", "password" -> "12345")
+            )
+        )
+      status(response) mustBe BAD_REQUEST
+      session(response).get(SessionKeys.token) mustBe empty
+    }
+
+    "redirect to login page given an invalid username and valid password from a new instance of controller" in {
```
Comment (tna-erasmos) [inline test/controllers/LoginControllerSpec.scala:88]: It's not a redirect, but rather a re-presentation of the same page.

File: test/controllers/LoginControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+          CSRFTokenHelper
+            .addCSRFToken(
+              FakeRequest(POST, "/login").withFormUrlEncodedBody("username" -> "123", "password" -> "1234")
+            )
+        )
+      status(response) mustBe BAD_REQUEST
+      session(response).get(SessionKeys.token) mustBe empty
+    }
+
+    "redirect to login page given an invalid username and password from a new instance of controller" in {
```
Comment (tna-erasmos) [inline test/controllers/LoginControllerSpec.scala:102]: It's not a redirect, but rather a re-presentation of the same page.

File: test/support/BaseControllerSpec.scala [hunk 8]
```diff
@@ -21,10 +21,18 @@
...
+import play.api.test.Helpers.{ contentAsString, defaultAwaitTimeout }
 import uk.gov.nationalarchives.omega.editorial.models.session.Session
 
+import scala.concurrent.Future
+
 class BaseControllerSpec extends UnitTest {
 
   val validSessionToken: String = Session.generateToken("1234")
   val invalidSessionToken: String = Session.generateToken("invalid-user")
+  def asDocument(html: String): Document = Jsoup.parse(html)
```
Comment (tna-erasmos) [inline test/support/BaseControllerSpec.scala:36]: You won't need these, as we're not supposed to assert contents in these controller specs.

**Review by tna-erasmos — COMMENTED** [review]
Minor changes suggested.

File: test/controllers/LoginControllerSpec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (userjd) [reply test/controllers/LoginControllerSpec.scala:47]: Removed

File: test/controllers/LoginControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+import uk.gov.nationalarchives.omega.editorial.controllers.{ LoginController, SessionKeys }
+import uk.gov.nationalarchives.omega.editorial.views.html.login
+
+class LoginControllerSpec extends BaseControllerSpec {
+
+  "LoginController GET" should {
+
+    "render the login page from a new instance of controller" in {
+
+      val response = getLoginController.view().apply(CSRFTokenHelper.addCSRFToken(FakeRequest(GET, "/login")))
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/controllers/LoginControllerSpec.scala:39]: Done

File: test/controllers/LoginControllerSpec.scala (hunk 2 above)
Comment (userjd) [reply test/controllers/LoginControllerSpec.scala:47]: Removed

File: test/support/BaseControllerSpec.scala (hunk 8 above)
Comment (userjd) [reply test/support/BaseControllerSpec.scala:36]: Removed

File: test/controllers/LoginControllerSpec.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/controllers/LoginControllerSpec.scala]: Confirmed.

File: test/controllers/LoginControllerSpec.scala (hunk 2 above)
Comment (tna-erasmos) [reply test/controllers/LoginControllerSpec.scala]: Confirmed.

File: test/controllers/LoginControllerSpec.scala (hunk 4 above)
Comment (tna-erasmos) [reply test/controllers/LoginControllerSpec.scala:71]: Confirmed.

**Pull request merged**